plugins/saveactions.c

```diff
@@ -304,6 +314,323 @@ static void instantsave_document_new_cb(GObject *obj, GeanyDocument *doc, gpoint
+	gtk_dialog_add_action_widget(GTK_DIALOG(dialog), button, GTK_RESPONSE_NO);
+	gtk_widget_show(button);
+
+	gtk_dialog_add_button(GTK_DIALOG(dialog), GTK_STOCK_SAVE, GTK_RESPONSE_YES);
+
+	gtk_dialog_set_default_response(GTK_DIALOG(dialog), GTK_RESPONSE_YES);
+	ret = gtk_dialog_run(GTK_DIALOG(dialog));
+
+	gtk_widget_destroy(dialog);
+
+	return ret;
+}
+
+
+static void show_unsaved_dialog_for_persistent_temp_files_tab_closing(
+	GeanyDocument *doc, const gchar *short_filename)
+{
+	gchar *msg, *old_file_path, *new_file_path, *locale_file_path;
+	const gchar *msg2;
+	gint response;
+
+	msg = g_strdup_printf(_("The file '%s' is not saved."), short_filename);
+	msg2 = _("Do you want to save it before closing?");
+
+	response = run_unsaved_dialog_for_persistent_temp_files_tab_closing(msg, msg2);
+	g_free(msg);
+
+	locale_file_path = g_strdup(doc->real_path);
+
+	switch (response)
+	{
+		case GTK_RESPONSE_YES:
+			old_file_path = g_strdup(doc->file_name);
+
+			if (dialogs_show_save_as())
+			{
+				new_file_path = doc->file_name;
+
+				if (! g_str_equal(old_file_path, new_file_path))
+				{
+				 	/* remove temp file if it was saved as some other file */
+					g_remove(locale_file_path);
+				}
+			}
+			else
+			{
+				plugin_idle_add(geany_plugin, open_document_once_idle, g_strdup(locale_file_path));
+			}
+
+			g_free(old_file_path);
+			break;
+
+		case GTK_RESPONSE_NO:
+			g_remove(locale_file_path);
+
+			ui_set_statusbar(TRUE, _("Temp file %s was deleted"), short_filename);
+			break;
+
+		case GTK_RESPONSE_CANCEL:
+		default:
+			plugin_idle_add(geany_plugin, open_document_once_idle, g_strdup(locale_file_path));
+			break;
+	}
+
+	g_free(locale_file_path);
+}
+
+
+static void persistent_temp_files_document_close_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	gchar *short_filename;
+
+	if (enable_persistent_temp_files && doc->real_path != NULL
+		&& is_temp_saved_file_doc(doc) && ! geany_is_closing_all_tabs())
+	{
+		short_filename = document_get_basename_for_display(doc, -1);
+
+		if (! document_is_empty(doc))
+		{
+			show_unsaved_dialog_for_persistent_temp_files_tab_closing(
+				doc,
+				short_filename
+			);
+		}
+		else
+		{
+			g_remove(doc->real_path);
+
+			ui_set_statusbar(TRUE, _("Empty temp file %s was deleted"), short_filename);
+		}
+
+		g_free(short_filename);
+	}
+}
+
+
+static void persistent_temp_files_document_before_save_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	if (enable_persistent_temp_files)
+	{
+		gchar *old_file_path, *new_file_path, *old_file_name;
+				GtkWidget *page_label = NULL;
+
+		get_current_tab_label(&page_label);
+
+		old_file_path = gtk_widget_get_tooltip_text(page_label);
+		new_file_path = DOC_FILENAME(doc);
+
+		if (old_file_path == NULL)
+		{
+			ui_set_statusbar(TRUE, _("plugin error: failed to delete initial temp file "
+				"('failed to get notebook tab label text')"));
+
+			return;
+		}
+
+		old_file_name = g_path_get_basename(old_file_path);
```

Comment:
I' just move `g_path_get_basename()` inside `is_temp_saved_file()` so it could be dropped here (and also inside ``is_temp_saved_file_doc()``)

Reply:
done